students/gregdevore/lesson09/donor_models.py

```diff
@@ -0,0 +1,203 @@
+
+    def lastDonation(self):
+        """
+        Return last donation for donor.
+        Returns $0 if donor has no donation history
+        """
+        return self.donations[-1] if self.donations else 0
+
+    def totalDonations(self):
+        """
+        Return total donations for donor
+        """
+        return sum(self.donations)
+
+    def numDonations(self):
+        """
+        Return number of donations made by donor
+        """
+        return len(self.donations)
+
+    def averageDonation(self):
+        """
+        Return the average donation made by donor
+        """
+        return self.totalDonations()/self.numDonations() if self.donations else 0
+
+    def generateEmail(self):
+        """
+        Generate email to donor
+        """
+        email_dict = {'donor_name':self.name,
+                      'donation_amount':self.lastDonation(),
+                      'total_amount':self.totalDonations()}
+
+        # Create formatted email that can be copied & pasted
+        email = ('\n'.join(['Dear {donor_name},','',
+        'Thank you for your generous donation of ${donation_amount:.2f}.',
+        'To date, you have donated a total of ${total_amount:.2f} to our charity.',
+        'Your contributions help new arrivals receive the highest quality care possible.',
+        'Please know that your donations make a world of difference!',
+        '','Sincerely,','The Good Place Team'])).format(**email_dict)
+
+        return(email)
+
+class DonorCollection():
+    """
+    Class representing a collection of Donor instances
+
+    Args:
+        donors (dict):
+            Dictionary of donor objects (key = donor name, value = donor object)
+
+    Properties:
+        donorNames (list):
+            List of donor names. Only getter functionality enabled.
+
+    Methods:
+        updateDonor(name, amount):
+            Creates a new donor with amount if not in collection, otherwise
+            adds donation to existing donor
+
+        getDonor(name):
+            Returns Donor instance (returns new Donor if not found)
+
+        generateReportData():
+            Generates report data in the form of a tuple for each donor.
+            The tuple contains the donors name, total donations, number of
+            donations, and average donation
+
+    """
+    def __init__(self):
+        """
+        Create new DonorCollection instance with empty donor dictionary
+        """
+        self.donors = {}
+
+    def updateDonor(self, name, amount):
+        """
+        Updates collection with new donation for donor
+
+        Args:
+            name (str):
+                Donor's name
+            amount (float):
+                Donation amount
+        """
+        # Create donor object if it doesn't exist, otherwise retrieve from collection
+        d = self.getDonor(name)
+        # Update donor object and replace in collection
+        d.addDonation(amount)
+        self.donors[name] = d
```

Comment:
remember that you can always shortcut since you're using temp var `d`:
 `self.donors[name] = self.get_donor(name).add_donation(amount)`